### ISSUE TITLE
Port Forwarding partially working

### ISSUE THREAD

**Issue by mikecoe** [Azure/aztk#434]
I have created a PySpark cluster inline with your documentation successfully. Once i SSH onto the cluster the port forwarding works but only for ports 8080 and 4040. I cant access 18080 and 8888 for the Jupyter notebook. Have you seen this before. I can run PySpark from the command line no problem

**Comment by timotheeguerin** [comment]
Did you include jupyter in your cluster configuration. It must either be installed as a custom script or if you use the latest from master as the new plugin system?

**Comment by mikecoe** [comment]
Thanks for the response, yes i have done it using a custom-script as per the document. This is my cluster.yaml and i have the jupyter.sh file in the custom-scripts directory.  How can i tell on my cluster if it has installed?


##  cluster settings

# id: <id of the cluster to be created, reccommended to specify with --id command line parameter>

# vm_size: <vm-size, see available options here: https://azure.microsoft.com/en-us/pricing/details/vir$
vm_size: standard_a2

# size: <number of dedicated nodes in the cluster, not that clusters must contain all dedicated or all$
size: 2

# size_low_pri: <number of low priority nodes in the cluster, mutually exclusive with size setting>


# username: <username for the linux user to be created> (optional)
username: spark

# docker_repo: <name of docker image repo (for more information, see https://github.com/Azure/aztk/blo$
docker_repo: aztk/python:spark2.2.0-python3.6.2-base

# # optional custom scripts to run on the Spark master, Spark worker or all nodes in the cluster
custom_scripts:
   - script: custom-scripts/jupyter.sh
     runOn: master
#   - script: <./relative/path/to/other/script.sh or ./relative/path/to/other/script/directory/>
#     runOn: <master/worker/all-nodes>

# To add your cluster to a virtual network provide the full arm resoruce id below
# subnet_id: /subscriptions/********-****-****-****-************/resourceGroups/********/providers/Mic$

# To define plugins
# plugins:
#   - name: rstudio_server
#     args:
#       version: 1.2.3

# wait: <true/false>
wait: false

**Comment by timotheeguerin** [comment]
Your config looks good, I would expect it to work, so something might have gone wrong. Did you try creating multiple clusters?
Could you check also which version did you checkout(stable, latest from master, old master, etc.)

For the custom scripts you have to check the docker.log file on the master node. If you are using BatchLabs you can go on the node and have a file explorer and the docker log file should be under `startup/wd`.
If you use plugins, each plugin has its own log file which make it easier to find problem
You can send the content of the docker.log, we'll try to figure out the problem

**Comment by paselem** [comment]
@mikecoe - one more thing, 18080 is the history server UI and will only work if you have enabled this history server (which we do not do by default). 

In your .aztk/spark-default.conf file you will need to uncomment the following two lines:
```sh
# spark.eventLog.enabled           true
# spark.eventLog.dir               hdfs://namenode:8021/directory
```
To get started with this history server, I would suggest setting the __spark.eventLog.dir__ to a local path on disk, for example: __file:///home/spark-current/eventLog__ or something like that. We will make sure we create the directory if it doesn't exist.

Regarding jupyter, we recently made a change to move that from a CustomScript to a Plugin. Can you update your cluster config as follows
```yaml

# custom_scripts:
    # - script: custom-scripts/jupyter.sh
    #  runOn: master

plugins:
  - name: jupyter
```

**Comment by mikecoe** [comment]
Thanks pablo, I now have Jupyter up and running

**Comment by paselem** [comment]
@mikecoe - Great! Will close out this issue.

@timotheeguerin - Can we please update the docs and remove the custom script going forward to avoid further confusion?

**Comment by mikecoe** [comment]
Hi pablo,

Please may I ask one more question.  I am connecting our cluster to to azure data lake and using Jupyter.  This is now working and I can run jobs against the data in the lake.

I am trying to determine the best way to persist the Jupyter notebooks when the cluster is deleted.  It seems a bit trickier if you are using the azure data lake?

What’s the recommended approach? Do I need to write a script to ssh on to the head node and copy the files somewhere?

Thanks
Mike

**Comment by paselem** [comment]
Hi @mikecoe - great question. Our recommended path for doing that is to use Azure Files and save off all of your notebooks there. That way as nodes come and go your files are always persisted. Note, that we do **not** recommend using Azure Files for large data sets since the throughput is not nearly as good as using WASB or ADL, but for smaller files like scripts, logs and notebooks it is usually a really nice option. Additional data can be found here: https://github.com/Azure/aztk/blob/master/docs/14-azure-files.md.

Here is a snippet of my cluster config where I mount my Azure Files
```yaml
azure_files:
    - storage_account_name: mystorageaccount
      storage_account_key: *********************************************************************************
      # Name of the file share in Azure Files
      file_share_path: notebooks
      # Mount point on the node in the cluster
      mount_path: /mnt/notebooks
```
And finally, this is what it looks like when I navigate to http://localhost:8888 (note the 'notebooks' directory is what has been added. It contains all my personal notebooks. We have also seen other users share this directory across a team of scientists to share notebooks and scripts).

![image](https://user-images.githubusercontent.com/11393045/37529425-56f927a4-28f4-11e8-95c8-15060d9583ed.png)

**Comment by mikecoe** [comment]
Hi pablo,

Thanks for this, I have this working now and it’s a great solution.

We are keen to use aztk and I’m very grateful for your help along the way

Thanks
Mike

**Comment by mikecoe** [comment]
Hi pablo,

Could I ask one final question? There is one more step in our journey.

We now have the ability to spin up clusters on demand, access the data lake and persist our Jupyter notebooks. We just need to schedule these notebooks to run daily. Again, I have read around options on this, but would really appreciate your recommendation on the subject?

Thanks again
Mike

**Comment by paselem** [comment]
Hi @mikecoe - yeah, that is a scenario we've heard of before as well. This comes down to two things:
1. The ability to schedule jobs
2. The ability to run notebooks as 'jobs'

Both of those are on our backlog, and not part of the current offering. That said, I think it would be reasonably easy to build a pipeline around AZTK using [Azure Functions](https://azure.microsoft.com/en-us/services/functions/). You can easily schedule functions on a timer or storage trigger, and then use that to have AZTK kick off a job which runs your notebook. Since we currently do not have the 'run notebook' support you could simply convert it using [nbconvert](https://github.com/jupyter/nbconvert) to a [python file](https://nbconvert.readthedocs.io/en/latest/customizing.html#Converting-a-notebook-to-an-(I)Python-script-and-printing-to-stdout) first and then run that.

In summary:
- Azure Function triggers (on a timer?)
- Azure Function Converts the notebook to a script
- Azure Function kicks off an AZTK job (note that you could manually manage the cluster yourself by doing aztk cluster create, aztk cluster submit and aztk cluster delete, but the __aztk job submit__ does all of those for you as a single command)
  - Your job would need to have the logic of persisting your outputs to ADL, which I assume it already does.

**Comment by mikecoe** [comment]
That’s great thanks pablo I shall try and set that up later today.

Another question I have relates to permissions to access ADL. From my Jupyter Notebook currently I can only read from my ADL, not write.  I’m a little unsure how I should be setting up these permissions so that I can write to ADL?

I tried to create a user on my aztk cluster with same credentials as my permanent cluster that accesses the ADL and then ssh on but that didn’t work.

I also tried to open up permissions on the directory I want to write too for everyone but that didn’t work either.  

Thanks
Mike

**Comment by paselem** [comment]
Hi @mikecoe - the only credentials that should matter when connecting to ADL is the Service Principal you created and added to your core-site.xml file. Take a look here: https://github.com/Azure/aztk/blob/master/aztk_cli/config/core-site.xml

I assume you've already done that though if you are able to read from the ADL source and registered the file system. If that is the case, then the next think you need to make sure of is that you've given sufficient permissions to the Service Principal you created in order to have write access to ADL.

Hopefully this helps!
